src/cli/bunx_command.zig

```diff
@@ -148,43 +149,101 @@ pub const BunxCommand = struct {
+                            if (positional.len == 1 and positional[0] == 'x')
+                                continue;
+                        }
+                    }
+
+                    if (package_name_for_update_request[0].len == 0 and positional.len > 0 and positional[0] != '-') {
+                        package_name_for_update_request[0] = positional;
+                        bin_name[0] = positional;
+                        continue;
+                    }
+
+                    try passthrough_list.append(positional);
+                }
+            }
+        } else {
+            // New behavior
+            // bun x cowsay hello // this is existing shortcut behavior
+            // bun x -- cowsay hello
+            // bun x --package=cowsay -- cowthink hello
+            // bunx -- cowsay hello
+            // bunx --package=cowsay -- cowthink hello
+            // bunx --package=cowsay -- cowthink --version
+
+            // current bug:
+            // when package is specified, the command that gets run is implicit
+            // but really the "what package should I run" logic should look like this:
+            // to_run = first item in passthrough
+            package_name_for_update_request[0] = cli.package;
+            var parsing_passthroughs = false;
+            var captured_bin_name = false;
 
             for (argv) |positional_| {
                 const positional = bun.span(positional_);
 
                 if (positional.len == 0) continue;
 
-                if (positional[0] != '-') {
-                    if (!found_subcommand_name) {
-                        found_subcommand_name = true;
-                        if (positional.len == 1 and positional[0] == 'x')
-                            continue;
-                    }
+                if (positional.len == 2 and positional[0] == '-' and positional[1] == '-') {
+                    parsing_passthroughs = true;
+                    continue;
                 }
 
-                if (!run_in_bun and !found_subcommand_name) {
-                    if (strings.eqlComptime(positional, "--bun")) {
-                        run_in_bun = true;
-                        continue;
-                    }
+                if (!parsing_passthroughs) {
+                    continue;
                 }
 
-                if (package_name_for_update_request[0].len == 0 and positional.len > 0 and positional[0] != '-') {
-                    package_name_for_update_request[0] = positional;
+                if (captured_bin_name == false) {
+                    bin_name[0] = positional;
+                    captured_bin_name = true;
                     continue;
                 }
 
-                try passthrough_list.append(positional);
+                if (captured_bin_name and parsing_passthroughs) {
+                    try passthrough_list.append(positional);
+                }
             }
         }
 
+        if (cli.package.len != 0 and bin_name[0].len == 0) {
```

Comment:
Specifying `--package` without separating `bunx` flags from `bin` with `--` is ambiguous and confusing, so it seems better to forbid that for now.